src/Transform/ONNX/DecomposeAfter.cpp

```diff
@@ -0,0 +1,438 @@
+    ArrayAttr outputShapeAttr) {
+  ONNXConvOpShapeHelper shapeHelper(op.getOperation(), {});
+  shapeHelper.computeShapeAndAssertOnFailure();
+  int inputRank = shapeHelper.getOutputDims().size();
+  SmallVector<int64_t, 4> inputShape;
+  for (int i = 0; i < inputRank; ++i) {
+    int64_t d = shapeHelper.getOutputDims()[i].isLiteral()
+                    ? shapeHelper.getOutputDims()[i].getLiteral()
+                    : ShapedType::kDynamic;
+    inputShape.emplace_back(d);
+  }
+  SmallVector<int64_t, 2> padSize;
+  int64_t attrSize = ArrayAttrSize(outputShapeAttr);
+  int64_t offset = inputRank - attrSize;
+  for (int i = 0; i < attrSize; ++i) {
+    IntegerAttr attr = outputShapeAttr.getValue()[i].cast<IntegerAttr>();
+    int64_t size = attr.getValue().getSExtValue() - inputShape[offset + i];
+    assert(size >= 0 && "Invalid output_shape attribute");
+    padSize.emplace_back(size);
+  }
+  Value paddedInput = emitPadsAxisEnd(
+      rewriter, loc, input, ArrayRef(inputShape), /*axis*/ 2, padSize[0]);
+  for (int i = 1; i < attrSize; ++i) {
+    ArrayRef<int64_t> paddedInputShape =
+        paddedInput.getType().cast<ShapedType>().getShape();
+    paddedInput = emitPadsAxisEnd(rewriter, loc, paddedInput, paddedInputShape,
+        /*axis*/ 2 + i, padSize[i]);
+  }
+  return paddedInput;
+}
+// ConvTransposeOp END
+
+} // namespace onnx_mlir
+
+namespace {
+/// Include the patterns defined in the Declarative Rewrite framework.
+#include "src/Transform/ONNX/ONNXDecomposeAfter.inc"
+
+struct DecomposeONNXToONNXAfterPass
+    : public PassWrapper<DecomposeONNXToONNXAfterPass,
+          OperationPass<func::FuncOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(DecomposeONNXToONNXAfterPass)
+
+  DecomposeONNXToONNXAfterPass(const std::string &target) {
+    this->target = target;
+  }
+  DecomposeONNXToONNXAfterPass(const DecomposeONNXToONNXAfterPass &pass)
+      : mlir::PassWrapper<DecomposeONNXToONNXAfterPass,
+            OperationPass<func::FuncOp>>() {
+    this->target = pass.target;
+  }
+
+  StringRef getArgument() const override { return "decompose-onnx-after"; }
+
+  StringRef getDescription() const override {
+    return "Decompose ONNX operations into composition of other ONNX "
+           "operations after shape inference.";
+  }
+
+  Option<std::string> target{*this, "target",
+      llvm::cl::desc("Target Dialect to decompose into"), ::llvm::cl::init("")};
+
+  void runOnOperation() final;
+
+  typedef PassWrapper<DecomposeONNXToONNXAfterPass, OperationPass<func::FuncOp>>
+      BaseType;
+};
+
+void DecomposeONNXToONNXAfterPass::runOnOperation() {
+  func::FuncOp function = getOperation();
+  MLIRContext *context = &getContext();
+
+  ConversionTarget target(getContext());
+  target.addLegalDialect<ONNXDialect, arith::ArithDialect, func::FuncDialect>();
+
+  // These ops will be decomposed into other ONNX ops. Hence, they will not be
+  // available after this pass.
+  target.addIllegalOp<ONNXConvTransposeOp>();
```

Comment:
Perhaps, `DynamicLegalOp` is better, saying the rewrite will be applied only some inputs are ranked.

Reply:
Done in original decompose pass. Thanks!